### PR TITLE
[http] Introduce two new config properties to modify http behavior an…

### PR DESCRIPTION
**Pull request by kumpera**
…d change the defaults.

Introduce http.timeout and http.certvalidation.

The first controls request timeout. Default value changes from 2 seconds to 30 seconds.
The second controls whether to check for server certificate errors. Default now is to always check.

There are few things to consider here.

- Whether we should be explicit about the default timeout or leave it up to the http client.
- If 30 seconds is a good or bad default

In addition to that, we should consider exposing the more http config knobs.

### CODE REVIEW

File: include/constants.h [hunk 1]
```diff
@@ -34,6 +34,8 @@ namespace reinforcement_learning {  namespace name {
       const char *const  QUEUE_MODE = "queue.mode";
       const char *const  INTERACTION_FILE_NAME = "interaction.file.name";
       const char *const  OBSERVATION_FILE_NAME = "observation.file.name";
+      const char *const  HTTP_CLIENT_DISABLE_CERT_VALIDATION  = "http.certvalidation";
```
Comment (lokitoth) [inline include/constants.h]: We should keep the negations of the constant name and the constant value similar. In other words, I would expect false to mean that we do cert validation when I read `HTTP_CLIENT_DISABLE_CERT_VALIDATION`, but the opposite when I read `http.certvalidation`.

File: include/constants.h [hunk 2]
```diff
@@ -34,6 +34,8 @@ namespace reinforcement_learning {  namespace name {
...
+      const char *const  HTTP_CLIENT_TIMEOUT   = "http.timeout"; //timeout is in seconds, default is 30
```
Comment (jackgerrits) [inline include/constants.h]: nit: comment style
```suggestion
      const char *const  HTTP_CLIENT_TIMEOUT   = "http.timeout"; // Timeout is in seconds, default is 30.
```

File: rlclientlib/utility/http_client.cc [hunk 3]
```diff
@@ -4,9 +4,9 @@
 #include <chrono>
 
 namespace reinforcement_learning {
-  http_client::http_client(const char* url)
+  http_client::http_client(const char* url, const utility::configuration &cfg)
```
Comment (jackgerrits) [inline rlclientlib/utility/http_client.cc]: nit: ref on type
```suggestion
  http_client::http_client(const char* url, const utility::configuration& cfg)
```

File: rlclientlib/utility/http_client.h [hunk 4]
```diff
@@ -19,7 +20,7 @@ namespace reinforcement_learning {
 
   class http_client : public i_http_client {
   public:
-    http_client(const char* url);
+    http_client(const char* url, const utility::configuration &cfg);
```
Comment (jackgerrits) [inline rlclientlib/utility/http_client.h]: nit: ref on type
```suggestion
    http_client(const char* url, const utility::configuration& cfg);
```

File: rlclientlib/utility/http_helper.cc [hunk 5]
```diff
@@ -1,15 +1,19 @@
 #include "http_helper.h"
+#include "constants.h"
 
 #include <chrono>
 
 namespace reinforcement_learning {
   namespace utility {
-    web::http::client::http_client_config get_http_config() {
+    web::http::client::http_client_config get_http_config(const utility::configuration &cfg) {
```
Comment (jackgerrits) [inline rlclientlib/utility/http_helper.cc]: nit: ref on typ
```suggestion
    web::http::client::http_client_config get_http_config(const utility::configuration& cfg) {
```

File: rlclientlib/utility/http_helper.cc [hunk 6]
```diff
@@ -1,15 +1,19 @@
...
       web::http::client::http_client_config config;
 
-      config.set_validate_certificates(false);
-      // Set a timeout for network requests to avoid hanging too long because of network issues.
-      config.set_timeout(std::chrono::seconds(2));
+      config.set_validate_certificates(cfg.get_bool(name::HTTP_CLIENT_DISABLE_CERT_VALIDATION, false));
+      int timeout = cfg.get_int(name::HTTP_CLIENT_TIMEOUT, 30);
```
Comment (jackgerrits) [inline rlclientlib/utility/http_helper.cc]:
```suggestion
      const auto timeout = cfg.get_int(name::HTTP_CLIENT_TIMEOUT, 30);
```

File: rlclientlib/utility/http_helper.cc [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
+	    //Valid values are 1-30
+      if (timeout < 1 || timeout > 30)
+		    timeout = 30;
```
Comment (jackgerrits) [inline rlclientlib/utility/http_helper.cc]: tabs and spaces are mixed here

File: rlclientlib/utility/http_helper.h [hunk 8]
```diff
@@ -1,8 +1,9 @@
 #pragma once
 #include <cpprest/http_client.h>
+#include "configuration.h"
 
 namespace reinforcement_learning { namespace utility {
 
-    web::http::client::http_client_config get_http_config();
+    web::http::client::http_client_config get_http_config(const utility::configuration &cfg);
```
Comment (jackgerrits) [inline rlclientlib/utility/http_helper.h]:
```suggestion
    web::http::client::http_client_config get_http_config(const utility::configuration& cfg);
```

File: rlclientlib/utility/http_helper.cc [hunk 9]
```diff
@@ -1,15 +1,19 @@
...
 namespace reinforcement_learning {
   namespace utility {
-    web::http::client::http_client_config get_http_config() {
+    web::http::client::http_client_config get_http_config(const utility::configuration &cfg) {
       web::http::client::http_client_config config;
 
-      config.set_validate_certificates(false);
-      // Set a timeout for network requests to avoid hanging too long because of network issues.
-      config.set_timeout(std::chrono::seconds(2));
+      config.set_validate_certificates(cfg.get_bool(name::HTTP_CLIENT_DISABLE_CERT_VALIDATION, false));
```
Comment (rajan-chari) [inline rlclientlib/utility/http_helper.cc]: Based on comments above, should this be HTTP_CLIENT_ENABLE_CERT_VALIDATION?

**Comment by rajan-chari** [comment]
Looks good.  Would like to see response to HTTP_CLIENT_DISABLE_CERT_VALIDATION question/comments before merging.

File: rlclientlib/utility/http_helper.cc (hunk 6 above)
Comment (kumpera) [reply rlclientlib/utility/http_helper.cc]: can't use const here as I change it down

File: rlclientlib/utility/http_helper.cc (hunk 7 above)
Comment (kumpera) [reply rlclientlib/utility/http_helper.cc]: 🤦‍♂️

File: rlclientlib/utility/http_helper.cc (hunk 9 above)
Comment (kumpera) [reply rlclientlib/utility/http_helper.cc]: No, it should be that, the logic in http_helper.cc is wrong

**Comment by kumpera** [comment]
The last commit addresses the issues raised here.

**Pull request merged**